vMenu/FunctionsController.cs

```diff
@@ -714,26 +712,44 @@ private async Task UpdateLocation()
             if (MainMenu.MiscSettingsMenu.ShowLocation)
             {
                 // Get the current location.
-                currentPos = GetEntityCoords(Game.PlayerPed.Handle, true);
+                var currentPos = GetEntityCoords(Game.PlayerPed.Handle, true);
+                var heading = Game.PlayerPed.Heading;
+                zoneDisplay = World.GetZoneLocalizedName(currentPos);
 
                 // Get the nearest vehicle node.
-                nodePos = currentPos;
+                var nodePos = Vector3.Zero;
                 GetNthClosestVehicleNode(currentPos.X, currentPos.Y, currentPos.Z, 0, ref nodePos, 0, 0, 0);
-                heading = Game.PlayerPed.Heading;
 
                 // Get the safezone size for x and y to be able to move with the minimap.
                 safeZoneSizeX = (1 / GetSafeZoneSize() / 3.0f) - 0.358f;
-                //safeZoneSizeY = GetSafeZoneSize() - 0.27f;
-                //safeZoneSizeY = (1 / GetSafeZoneSize() / 3.6f) - 0.27f;
 
                 // Get the cross road.
-                var p1 = (uint)1; // unused
-                crossing = (uint)1;
-                GetStreetNameAtCoord(currentPos.X, currentPos.Y, currentPos.Z, ref p1, ref crossing);
-                crossingName = GetStreetNameFromHashKey(crossing);
+                uint mainSt = 0, crossSt = 0;
+                GetStreetNameAtCoord(currentPos.X, currentPos.Y, currentPos.Z, ref mainSt, ref crossSt);
+                var mainName = GetStreetNameFromHashKey(mainSt);
+                var crossName = GetStreetNameFromHashKey(crossSt);
 
                 // Set the suffix for the road name to the corssing name, or to an empty string if there's no crossing.
-                suffix = (crossingName != "" && crossingName != "NULL" && crossingName != null) ? "~t~ / " + crossingName : "";
+                var prefix = currentPos.DistanceToSquared(nodePos) > 1400f ? "~m~Near ~s~" : "~s~";
+                var suffix = crossSt != 0 ? "~t~ / " + crossName : "";
+                streetDisplay = prefix + mainName + suffix;
+
+                if (heading > 320 || heading < 45) // North
+                {
+                    headingDisplay = "N";
+                }
+                else if (heading >= 45 && heading <= 135) // West
+                {
+                    headingDisplay = "W";
+                }
+                else if (heading > 135 && heading < 225) // South
+                {
+                    headingDisplay = "S";
+                }
+                else // East
+                {
+                    headingDisplay = "E";
+                }
 
                 await Delay(200);
```

Comment:
Have you experimented with increasing this value to `500` or even `1000`? If we're getting `0.3ms` decrease with only a `200` delay, it would be interesting to see if we could squeeze a bit more out of it.

Realistically, `200` is still a very short delay, the player isn't going to be changing their position or heading fast enough for `200` to be worthwhile. I suspect something around the `750` mark might be better, assuming it has a positive impact on the performance.

Reply:
The issue of performance was almost completely the `GetStreetNameAtCoord` and getting the names of the street and cross-street. Heading checks would be negligible, however I moved them up for continuity reasons (so they update when streets update). I think every `200ms` is a good metric for streets, cause people can move pretty fast in NoClip and such.